cli/src/action/certs.rs

```diff
@@ -0,0 +1,389 @@
+use openssl::error::ErrorStack;
+use openssl::hash::MessageDigest;
+use openssl::pkey::{PKey, PKeyRef, Private};
+use openssl::rsa::Rsa;
+use openssl::x509::extension::{BasicConstraints, ExtendedKeyUsage, KeyUsage};
+use openssl::x509::{X509NameBuilder, X509Ref, X509};
+
+use crate::error::CliError;
+
+use super::{chown, Action};
+
+pub struct CertGenAction;
+
+const DEFAULT_CERT_DIR: &str = "/etc/splinter/certs/";
+const CERT_DIR_ENV: &str = "SPLINTER_CERT_DIR";
+
+const CLIENT_CERT: &str = "client.crt";
+const CLIENT_KEY: &str = "client.key";
+const SERVER_CERT: &str = "server.crt";
+const SERVER_KEY: &str = "server.key";
+
+impl Action for CertGenAction {
+    fn run<'a>(&mut self, arg_matches: Option<&ArgMatches<'a>>) -> Result<(), CliError> {
+        let args = arg_matches.ok_or_else(|| CliError::RequiresArgs)?;
+
+        let common_name = args
+            .value_of("common_name")
+            .unwrap_or("localhost")
+            .to_string();
+
+        let cert_dir_string = args
+            .value_of("cert_dir")
+            .map(ToOwned::to_owned)
+            .or_else(|| env::var(CERT_DIR_ENV).ok())
+            .or_else(|| Some(DEFAULT_CERT_DIR.to_string()))
+            .unwrap();
+
+        let cert_dir = Path::new(&cert_dir_string);
+
+        // Check if the provided cert directory exists
+        if !cert_dir.is_dir() {
+            return Err(CliError::ActionError(format!(
+                "Cert directory does not exist: {}",
+                cert_dir.display()
+            )));
+        }
+
+        let client_cert_path = cert_dir.join(CLIENT_CERT);
+        let server_cert_path = cert_dir.join(SERVER_CERT);
+
+        let private_cert_dir = cert_dir.join("private/");
+
+        // Check if the provided private key directory for the certs exists, if not create it
+        if !private_cert_dir.is_dir() {
+            fs::create_dir_all(private_cert_dir.clone()).map_err(|err| {
+                CliError::ActionError(format!("Unable to create private directory {}", err))
+            })?
+        }
+
+        let client_key_path = private_cert_dir.join(CLIENT_KEY);
+        let server_key_path = private_cert_dir.join(SERVER_KEY);
+
+        // if check
+        if args.is_present("skip") {
+            if client_cert_path.exists()
+                && client_key_path.exists()
+                && server_cert_path.exists()
+                && server_key_path.exists()
+            {
+                println!("Certificates and keys exist");
+            } else if client_cert_path.exists()
+                || client_key_path.exists()
+                || server_cert_path.exists()
+                || server_key_path.exists()
+            {
+                // this will always be an error condition
+                return Err(CliError::ActionError(
+                    "Some but not all certificates and keys already exist".into(),
+                ));
+            } else {
+                println!("Certificates and keys do not exist");
```

Comment:
This should be more explicit about what does not exist (paths, etc.).

Do these really get turned off if you give -q? Maybe consider doing these with logging and configure the logger at the top-level, instead of passing a quiet boolean through the code.

Reply:
done